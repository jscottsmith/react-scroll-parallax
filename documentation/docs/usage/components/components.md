# Components

- [`<Parallax>`](/docs/usage/components/parallax-component)
- [`<ParallaxBanner>`](/docs/usage/components/parallax-banner-component)
- [`<ParallaxProvider>`](/docs/usage/components/parallax-provider)
