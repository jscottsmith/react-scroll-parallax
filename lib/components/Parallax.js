"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactObserved = require("react-observed");

var _reactObserved2 = _interopRequireDefault(_reactObserved);

var _propValidation = require("../utils/propValidation");

var _index = require("./index.js");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Parallax = function (_Component) {
    _inherits(Parallax, _Component);

    function Parallax() {
        _classCallCheck(this, Parallax);

        return _possibleConstructorReturn(this, (Parallax.__proto__ || Object.getPrototypeOf(Parallax)).apply(this, arguments));
    }

    _createClass(Parallax, [{
        key: "getWrapperClass",
        value: function getWrapperClass(isInView) {
            var className = this.props.className;

            var cx = "parallax-wrapper";
            cx = isInView ? cx + " is-in-view" : cx;
            cx = className ? cx + " " + className : cx;

            return cx;
        }
    }, {
        key: "render",
        value: function render() {
            var _this2 = this;

            // NOTE: Each component in this tree provides a render
            // callback with necessary params to provide as props
            // for the child component
            //
            // It's not as bad as it looks. ;)
            //
            // <Observed> => ({ isInView, mapRef })
            // ↓
            // <ScrollPosition> => ({ scrollY })
            // ↓
            // <ViewportProgress> => ({ progress, mapRef, updateAttributeCache })
            // ↓
            // <ScrollEffects> => ({ style })
            // ↓
            // markup

            var _props = this.props,
                children = _props.children,
                x = _props.x,
                y = _props.y,
                scale = _props.scale,
                opacity = _props.opacity,
                observerOptions = _props.observerOptions,
                elementStyle = _props.elementStyle;

            // if child is a function, call it with updateAttributeCache

            var isFunc = children instanceof Function;

            // prettier-ignore
            return _react2.default.createElement(
                _reactObserved2.default,
                { initialViewState: true, options: observerOptions },
                function (_ref) {
                    var isInView = _ref.isInView,
                        observedRef = _ref.mapRef;
                    return _react2.default.createElement(
                        _index.ScrollPosition,
                        { isInView: isInView },
                        function (_ref2) {
                            var scrollY = _ref2.scrollY;
                            return _react2.default.createElement(
                                _index.ViewportProgress,
                                { isInView: isInView, scrollY: scrollY },
                                function (_ref3) {
                                    var progress = _ref3.progress,
                                        updateAttributeCache = _ref3.updateAttributeCache,
                                        viewportRef = _ref3.mapRef;
                                    return _react2.default.createElement(
                                        _index.ScrollEffects,
                                        { progress: progress, x: x, y: y, scale: scale, opacity: opacity },
                                        function (_ref4) {
                                            var style = _ref4.style;
                                            return _react2.default.createElement(
                                                "div",
                                                { className: _this2.getWrapperClass(isInView) },
                                                _react2.default.createElement(
                                                    _index.Bounds,
                                                    {
                                                        refCallbacks: [observedRef, viewportRef],
                                                        updateAttributeCache: updateAttributeCache,
                                                        scale: scale,
                                                        x: x,
                                                        y: y
                                                    },
                                                    _react2.default.createElement(
                                                        "div",
                                                        _defineProperty({ style: style, className: "parallax-element" }, "style", elementStyle),
                                                        isFunc ? children({ updateAttributeCache: updateAttributeCache, progress: progress, isInView: isInView }) : children
                                                    )
                                                )
                                            );
                                        }
                                    );
                                }
                            );
                        }
                    );
                }
            );
            // prettier-ignore
        }
    }]);

    return Parallax;
}(_react.Component);

Parallax.defaultProps = {
    x: [0, 0],
    y: [0, 0],
    scale: [1, 1],
    opacity: [1, 1]
};
Parallax.propTypes = {
    children: _propTypes2.default.oneOfType([_propTypes2.default.element.isRequired, _propTypes2.default.func.isRequired]),
    className: _propTypes2.default.string,
    x: _propValidation.validateOffsets,
    y: _propValidation.validateOffsets,
    scale: _propValidation.validateScale,
    opacity: _propValidation.validateOpacity,
    observerOptions: _propTypes2.default.object,
    elementStyle: _propTypes2.default.object
    // @TODO: these should also be available:
    // rotation
    // tag/element name?
};
exports.default = Parallax;
module.exports = exports["default"];