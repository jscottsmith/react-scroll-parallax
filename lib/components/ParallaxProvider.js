'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactScrollParallax = require('react-scroll-parallax');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ParallaxProvider = function (_Component) {
    _inherits(ParallaxProvider, _Component);

    function ParallaxProvider() {
        _classCallCheck(this, ParallaxProvider);

        return _possibleConstructorReturn(this, (ParallaxProvider.__proto__ || Object.getPrototypeOf(ParallaxProvider)).apply(this, arguments));
    }

    _createClass(ParallaxProvider, [{
        key: 'getChildContext',
        value: function getChildContext() {
            // Passes down the reference to the controller
            var scrollController = this.scrollController,
                resizeController = this.resizeController;

            return { scrollController: scrollController, resizeController: resizeController };
        }
    }, {
        key: 'componentWillMount',
        value: function componentWillMount() {
            // Don't initialize on the server
            var hasWindow = typeof window !== 'undefined';

            if (hasWindow) {
                // Must not be the server so kick it off...
                this.scrollController = _reactScrollParallax.ScrollController.init();
                this.resizeController = _reactScrollParallax.ResizeController.init();
            }
        }
    }, {
        key: 'componentWillUnmount',
        value: function componentWillUnmount() {
            // Remove scroll and resize listener if the provider is unmounted

            if (this.scrollController) {
                this.scrollController.destroy();
            }

            if (this.resizeController) {
                this.resizeController.destroy();
            }
        }
    }, {
        key: 'render',
        value: function render() {
            var children = this.props.children;


            return children;
        }
    }]);

    return ParallaxProvider;
}(_react.Component);

ParallaxProvider.propTypes = {
    children: _propTypes2.default.node.isRequired
};
ParallaxProvider.childContextTypes = {
    scrollController: _propTypes2.default.object,
    resizeController: _propTypes2.default.object
};
exports.default = ParallaxProvider;
module.exports = exports['default'];