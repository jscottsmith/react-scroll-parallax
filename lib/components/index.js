'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ViewportProgress = exports.ParallaxProvider = exports.ScrollPosition = exports.ScrollEffects = exports.Parallax = exports.Bounds = undefined;

var _Bounds2 = require('./Bounds.js');

var _Bounds3 = _interopRequireDefault(_Bounds2);

var _Parallax2 = require('./Parallax.js');

var _Parallax3 = _interopRequireDefault(_Parallax2);

var _ScrollEffects2 = require('./ScrollEffects.js');

var _ScrollEffects3 = _interopRequireDefault(_ScrollEffects2);

var _ScrollPosition2 = require('./ScrollPosition.js');

var _ScrollPosition3 = _interopRequireDefault(_ScrollPosition2);

var _ParallaxProvider2 = require('./ParallaxProvider.js');

var _ParallaxProvider3 = _interopRequireDefault(_ParallaxProvider2);

var _ViewportProgress2 = require('./ViewportProgress.js');

var _ViewportProgress3 = _interopRequireDefault(_ViewportProgress2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.Bounds = _Bounds3.default; // Components

exports.Parallax = _Parallax3.default;
exports.ScrollEffects = _ScrollEffects3.default;
exports.ScrollPosition = _ScrollPosition3.default;
exports.ParallaxProvider = _ParallaxProvider3.default;
exports.ViewportProgress = _ViewportProgress3.default;