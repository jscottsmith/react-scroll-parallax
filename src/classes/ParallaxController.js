import { testForPassiveScroll } from '../utils';
import { resetStyles } from '../helpers';
import { View } from './View';
import { Scroll } from './Scroll';
import { Element } from './Element';
import { VERTICAL } from '../constants';

/**
 * -------------------------------------------------------
 * Parallax Controller
 * -------------------------------------------------------
 *
 * The global controller for setting up window scroll/resize
 * listeners, managing and caching parallax element positions,
 * determining which elements are inside the viewport based on
 * scroll position, and then updating parallax element styles
 * based on x/y offsets and current scroll position.
 *
 */
function ParallaxController({ scrollAxis = VERTICAL, scrollContainer }) {
    // All parallax elements to be updated
    let elements = [];

    let hasScrollContainer = !!scrollContainer;
    let viewEl = scrollContainer || window;

    // Scroll and View
    const x = hasScrollContainer ? viewEl.scrollLeft : window.pageXOffset;
    const y = hasScrollContainer ? viewEl.scrollTop : window.pageYOffset;
    const scroll = new Scroll(x, y);
    let view = new View({ width: 0, height: 0, scrollContainer });

    // Ticking
    let ticking = false;

    // Passive support
    const supportsPassive = testForPassiveScroll();

    function _addListeners(el) {
        el.addEventListener(
            'scroll',
            _handleScroll,
            supportsPassive ? { passive: true } : false
        );
        window.addEventListener('resize', _handleResize, false);
    }

    function _removeListeners(el) {
        el.removeEventListener(
            'scroll',
            _handleScroll,
            supportsPassive ? { passive: true } : false
        );
        window.removeEventListener('resize', _handleResize, false);
    }

    _addListeners(viewEl);
    _setViewSize();

    /**
     * Window scroll handler sets scroll position
     * and then calls '_updateAllElements()'.
     */
    function _handleScroll() {
        // Save current scroll
        // Supports IE 9 and up.
        const nx = hasScrollContainer ? viewEl.scrollLeft : window.pageXOffset;
        const ny = hasScrollContainer ? viewEl.scrollTop : window.pageYOffset;
        scroll.setScroll(nx, ny);

        // Only called if the last animation request has been
        // completed and there are parallax elements to update
        if (!ticking && elements.length > 0) {
            ticking = true;
            window.requestAnimationFrame(_updateAllElements);
        }
    }

    /**
     * Window resize handler. Sets the new window inner height
     * then updates parallax element attributes and positions.
     */
    function _handleResize() {
        _setViewSize();
        _updateAllElements({ updateCache: true });
    }

    /**
     * Update element positions.
     * Determines if the element is in view based on the cached
     * attributes, if so set the elements parallax styles.
     */
    function _updateAllElements({ updateCache } = {}) {
        if (elements) {
            elements.forEach(element => {
                _updateElementPosition(element);
                if (updateCache) {
                    element.setCachedAttributes(view, scroll);
                }
            });
        }
        // reset ticking so more animations can be called
        ticking = false;
    }

    /**
     * Update element positions.
     * Determines if the element is in view based on the cached
     * attributes, if so set the elements parallax styles.
     */
    function _updateElementPosition(element) {
        if (element.props.disabled) return;
        element.updatePosition(view, scroll);
    }

    /**
     * Cache the window height.
     */
    function _setViewSize() {
        if (hasScrollContainer) {
            const width = viewEl.offsetWidth;
            const height = viewEl.offsetHeight;
            return view.setSize(width, height);
        }

        const html = document.documentElement;
        const width = window.innerWidth || html.clientWidth;
        const height = window.innerHeight || html.clientHeight;

        return view.setSize(width, height);
    }

    /**
     * -------------------------------------------------------
     * Public methods
     * -------------------------------------------------------
     */

    /**
     * Gets the parallax elements in the controller
     * @return {array} parallax elements
     */
    this.getElements = function() {
        return elements;
    };

    /**
     * Creates a new parallax element object with new id
     * and options to store in the 'elements' array.
     * @param {object} options
     * @return {object} element
     */
    this.createElement = function(options) {
        const newElement = new Element({ ...options, scrollAxis });
        newElement.setCachedAttributes(view, scroll);
        elements = elements ? [...elements, newElement] : [newElement];
        _updateElementPosition(newElement);
        return newElement;
    };

    /**
     * Remove an element by id
     * @param {object} element
     */
    this.removeElementById = function(id) {
        if (!elements) return;
        elements = elements.filter(el => el.id !== id);
    };

    /**
     * Updates an existing parallax element object with new options.
     * @param {object} element
     * @param {object} options
     */
    this.updateElementPropsById = function(id, props) {
        if (elements) {
            elements = elements.map(el => {
                if (el.id === id) {
                    return el.updateProps(props);
                }
                return el;
            });
        }

        this.update();
    };

    /**
     * Remove element styles.
     * @param {object} element
     */
    this.resetElementStyles = function(element) {
        resetStyles(element);
    };

    /**
     * Updates all parallax element attributes and positions.
     */
    this.update = function() {
        _setViewSize();
        _updateAllElements({ updateCache: true });
    };

    this.updateScrollContainer = function(el) {
        // remove existing listeners with current el first
        _removeListeners(viewEl);

        viewEl = el;
        hasScrollContainer = !!el;
        view = new View({ width: 0, height: 0, scrollContainer: el });
        _setViewSize();
        _addListeners(viewEl);
        _updateAllElements({ updateCache: true });
    };

    /**
     * Removes listeners, reset all styles then nullifies the global ParallaxController.
     */
    this.destroy = function() {
        _removeListeners(viewEl);
        if (elements) {
            elements.forEach(element => resetStyles(element));
        }
        elements = undefined;
    };
}

/**
 * Static method to instantiate the ParallaxController.
 * @returns {Object} ParallaxController
 */
ParallaxController.init = function(options) {
    const hasWindow = typeof window !== 'undefined';

    if (!hasWindow) {
        throw new Error(
            'Looks like ParallaxController.init() was called on the server. This method must be called on the client.'
        );
    }

    return new ParallaxController(options);
};

export default ParallaxController;
