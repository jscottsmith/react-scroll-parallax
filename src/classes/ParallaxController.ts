import { resetStyles } from '../helpers/elementStyles';
import { View } from './View';
import { Scroll } from './Scroll';
import { Element } from './Element';
import { VERTICAL } from '../constants';
import { testForPassiveScroll } from '../utils/testForPassiveScroll';

export type ViewElement = HTMLElement | Window;

/**
 * -------------------------------------------------------
 * Parallax Controller
 * -------------------------------------------------------
 *
 * The global controller for setting up window scroll/resize
 * listeners, managing and caching parallax element positions,
 * determining which elements are inside the viewport based on
 * scroll position, and then updating parallax element styles
 * based on x/y offsets and current scroll position.
 *
 */

export type ParallaxControllerOptions = {
  scrollAxis?: 'vertical' | 'horizontal';
  scrollContainer?: HTMLElement;
};

export type ParallaxElementProperties = {
  disabled?: boolean;
  x0: string | number;
  x1: string | number;
  y0: string | number;
  y1: string | number;
};

export type CreateElementOptions = {
  elInner?: HTMLElement;
  elOuter?: HTMLElement;
  props: ParallaxElementProperties;
};

export class ParallaxController {
  elements: Element[];
  scrollAxis: 'vertical' | 'horizontal';
  viewEl: ViewElement;
  scroll: Scroll;
  view: View;
  _hasScrollContainer: boolean;
  _ticking: boolean;
  _supportsPassive: boolean;

  /**
   * Static method to instantiate the ParallaxController.
   * @returns {Class} ParallaxController
   */
  static init(options: ParallaxControllerOptions): ParallaxController {
    const hasWindow = typeof window !== 'undefined';

    if (!hasWindow) {
      throw new Error(
        'Looks like ParallaxController.init() was called on the server. This method must be called on the client.'
      );
    }

    return new ParallaxController(options);
  }

  constructor({
    scrollAxis = VERTICAL,
    scrollContainer,
  }: ParallaxControllerOptions) {
    this.scrollAxis = scrollAxis;
    // All parallax elements to be updated
    this.elements = [];

    this._hasScrollContainer = !!scrollContainer;
    this.viewEl = scrollContainer ?? window;

    // Scroll and View
    const x = this._hasScrollContainer
      ? // @ts-expect-error
        this.viewEl.scrollLeft
      : window.pageXOffset;
    const y = this._hasScrollContainer
      ? // @ts-expect-error
        this.viewEl.scrollTop
      : window.pageYOffset;

    this.scroll = new Scroll(x, y);
    this.view = new View({
      width: 0,
      height: 0,
      scrollContainer: this._hasScrollContainer ? scrollContainer : undefined,
    });

    // Ticking
    this._ticking = false;

    // Passive support
    this._supportsPassive = testForPassiveScroll();

    // Bind methods to class
    this._bindAllMethods();
    this._addListeners(this.viewEl);
    this._setViewSize();
  }

  private _bindAllMethods() {
    [
      '_addListeners',
      '_removeListeners',
      '_handleScroll',
      '_handleResize',
      '_updateAllElements',
      '_updateElementPosition',
      '_setViewSize',
      'getElements',
      'createElement',
      'removeElementById',
      'updateElementPropsById',
      'resetElementStyles',
      'update',
      'updateScrollContainer',
      'destroy',
    ].forEach((method: string) => {
      // @ts-expect-error
      this[method] = this[method].bind(this);
    });
  }

  private _addListeners(el: ViewElement) {
    el.addEventListener(
      'scroll',
      this._handleScroll,
      this._supportsPassive ? { passive: true } : false
    );
    window.addEventListener('resize', this._handleResize, false);
  }

  private _removeListeners(el: ViewElement) {
    el.removeEventListener('scroll', this._handleScroll, false);
    window.removeEventListener('resize', this._handleResize, false);
  }

  /**
   * Window scroll handler sets scroll position
   * and then calls '_updateAllElements()'.
   */
  private _handleScroll() {
    // Save current scroll
    // Supports IE 9 and up.
    const nx = this._hasScrollContainer
      ? // @ts-expect-error
        this.viewEl.scrollLeft
      : window.pageXOffset;
    const ny = this._hasScrollContainer
      ? // @ts-expect-error
        this.viewEl.scrollTop
      : window.pageYOffset;
    this.scroll.setScroll(nx, ny);

    // Only called if the last animation request has been
    // completed and there are parallax elements to update
    if (!this._ticking && this.elements.length > 0) {
      this._ticking = true;
      // @ts-ignore
      window.requestAnimationFrame(this._updateAllElements);
    }
  }

  /**
   * Window resize handler. Sets the new window inner height
   * then updates parallax element attributes and positions.
   */
  private _handleResize() {
    this._setViewSize();
    this._updateAllElements({ updateCache: true });
  }

  /**
   * Update element positions.
   * Determines if the element is in view based on the cached
   * attributes, if so set the elements parallax styles.
   */
  private _updateAllElements({ updateCache }: { updateCache?: boolean } = {}) {
    if (this.elements) {
      this.elements.forEach((element) => {
        this._updateElementPosition(element);
        if (updateCache) {
          element.setCachedAttributes(this.view, this.scroll);
        }
      });
    }
    // reset ticking so more animations can be called
    this._ticking = false;
  }

  /**
   * Update element positions.
   * Determines if the element is in view based on the cached
   * attributes, if so set the elements parallax styles.
   */
  private _updateElementPosition(element: Element) {
    if (element.props.disabled) return;
    element.updatePosition(this.view, this.scroll);
  }

  /**
   * Cache the window width/height.
   */
  private _setViewSize() {
    if (this._hasScrollContainer) {
      // @ts-ignore
      const width = this.viewEl.offsetWidth;
      // @ts-ignore
      const height = this.viewEl.offsetHeight;
      return this.view.setSize(width, height);
    }

    const html = document.documentElement;
    const width = window.innerWidth || html.clientWidth;
    const height = window.innerHeight || html.clientHeight;

    return this.view.setSize(width, height);
  }

  /**
   * -------------------------------------------------------
   * Public methods
   * -------------------------------------------------------
   */

  /**
   * Gets the parallax elements in the controller
   * @return {array} parallax elements
   */
  getElements(): Element[] {
    return this.elements;
  }

  /**
   * Creates a new parallax element object with new id
   * and options to store in the 'elements' array.
   * @param {object} options
   * @return {object} element
   */
  createElement(options: CreateElementOptions): Element {
    const newElement = new Element({ ...options, scrollAxis: this.scrollAxis });
    newElement.setCachedAttributes(this.view, this.scroll);
    this.elements = this.elements
      ? [...this.elements, newElement]
      : [newElement];
    this._updateElementPosition(newElement);
    return newElement;
  }

  /**
   * Remove an element by id
   * @param {object} element
   */
  removeElementById(id: number) {
    if (!this.elements) return;
    this.elements = this.elements.filter((el) => el.id !== id);
  }

  /**
   * Updates an existing parallax element object with new options.
   * @param {object} element
   * @param {object} options
   */
  updateElementPropsById(id: number, props: ParallaxElementProperties): void {
    if (this.elements) {
      this.elements = this.elements.map((el) => {
        if (el.id === id) {
          return el.updateProps(props);
        }
        return el;
      });
    }

    this.update();
  }

  /**
   * Remove element styles.
   * @param {object} element
   */
  resetElementStyles(element: Element) {
    resetStyles(element);
  }

  /**
   * Updates all parallax element attributes and positions.
   */
  update() {
    this._setViewSize();
    this._updateAllElements({ updateCache: true });
  }

  updateScrollContainer(el: HTMLElement) {
    // remove existing listeners with current el first
    this._removeListeners(this.viewEl);

    this.viewEl = el;
    this._hasScrollContainer = !!el;
    this.view = new View({ width: 0, height: 0, scrollContainer: el });
    this._setViewSize();
    this._addListeners(this.viewEl);
    this._updateAllElements({ updateCache: true });
  }

  /**
   * Removes listeners, reset all styles then nullifies the global ParallaxController.
   */
  destroy() {
    this._removeListeners(this.viewEl);
    if (this.elements) {
      this.elements.forEach((element) => resetStyles(element));
    }
    // @ts-expect-error
    this.elements = undefined;
  }
}
