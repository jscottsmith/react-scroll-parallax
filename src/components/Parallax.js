import React, { Component } from "react";
import PropTypes from "prop-types";
import Observed from "react-observed";
import {
    validateOffsets,
    validateScale,
    validateOpacity
} from "../utils/propValidation";
import {
    Bounds,
    ScrollEffects,
    ScrollPosition,
    ViewportProgress
} from "./index.js";

class Parallax extends Component {
    static defaultProps = {
        x: [0, 0],
        y: [0, 0],
        scale: [1, 1],
        opacity: [1, 1]
    };

    static propTypes = {
        children: PropTypes.oneOfType([
            PropTypes.element.isRequired,
            PropTypes.func.isRequired
        ]),
        className: PropTypes.string,
        x: validateOffsets,
        y: validateOffsets,
        scale: validateScale,
        opacity: validateOpacity,
        observerOptions: PropTypes.object,
        elementStyle: PropTypes.object
        // @TODO: these should also be available:
        // rotation
        // tag/element name?
    };

    getWrapperClass(isInView) {
        const { className } = this.props;
        let cx = "parallax-wrapper";
        cx = isInView ? `${cx} is-in-view` : cx;
        cx = className ? `${cx} ${className}` : cx;

        return cx;
    }

    render() {
        // NOTE: Each component in this tree provides a render
        // callback with necessary params to provide as props
        // for the child component
        //
        // It's not as bad as it looks. ;)
        //
        // <Observed> => ({ isInView, mapRef })
        // ↓
        // <ScrollPosition> => ({ scrollY })
        // ↓
        // <ViewportProgress> => ({ progress, mapRef, updateAttributeCache })
        // ↓
        // <ScrollEffects> => ({ style })
        // ↓
        // markup

        const {
            children,
            x,
            y,
            scale,
            opacity,
            observerOptions,
            elementStyle
        } = this.props;

        // if child is a function, call it with updateAttributeCache
        const isFunc = children instanceof Function;

        // prettier-ignore
        return (
            <Observed initialViewState options={observerOptions}>
                {({ isInView, mapRef: observedRef }) => (
                    <ScrollPosition isInView={isInView}>
                        {({ scrollY }) => (
                            <ViewportProgress isInView={isInView} scrollY={scrollY}>
                                {({ progress, updateAttributeCache, mapRef: viewportRef }) => (
                                    <ScrollEffects progress={progress} x={x} y={y} scale={scale} opacity={opacity}>
                                        {({ style }) => (
                                            <div className={this.getWrapperClass(isInView)}>
                                                <Bounds
                                                    refCallbacks={[observedRef, viewportRef]}
                                                    updateAttributeCache={updateAttributeCache}
                                                    scale={scale}
                                                    x={x}
                                                    y={y}
                                                >
                                                    <div style={style} className="parallax-element" style={elementStyle}>
                                                        {isFunc ? children({ updateAttributeCache, progress, isInView }) : children}
                                                    </div>
                                                </Bounds>                                                
                                            </div>
                                        )}
                                    </ScrollEffects>
                                )}
                            </ViewportProgress>
                        )}
                    </ScrollPosition>
                )}
            </Observed>
        );
        // prettier-ignore
    }
}

export default Parallax;
