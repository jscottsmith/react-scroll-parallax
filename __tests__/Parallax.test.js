import React from 'react';
import { render } from '@testing-library/react';
import { Parallax } from '../src/components/Parallax';
import { ParallaxProvider } from '../src/components/ParallaxProvider';
import { ParallaxController } from '../src/classes/ParallaxController';
import { VERTICAL } from '../src/constants';

import { MockProvider } from './testUtils/MockProvider';
import expectRenderError from './testUtils/expectRenderError';
import createNodeMock from './testUtils/createNodeMock';

const consoleLog = global.console.log;

describe('Expect the <Parallax> component', () => {
  const preventError = (e) => e.preventDefault();

  beforeEach(() => {
    window.addEventListener('error', preventError);
  });

  afterEach(() => {
    global.console.log = consoleLog;
    global.ParallaxController = undefined;

    window.removeEventListener('error', preventError);
  });

  it('to render correctly', () => {
    const { asFragment } = render(
      <ParallaxProvider>
        <Parallax
          className="class-foo"
          disabled={false}
          x={[-100, 100]}
          y={['-100%', '100%']}
          styleOuter={{
            border: 'solid red 2px',
          }}
          styleInner={{
            border: 'solid blue 2px',
          }}
          tagOuter="figure"
          tagInner="div"
        >
          <div className="foo" />
        </Parallax>
      </ParallaxProvider>,
      {
        createNodeMock,
      }
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('to throw if the ParallaxController is not available', () => {
    expectRenderError(
      <Parallax>
        <div />
      </Parallax>,
      'Could not find `react-scroll-parallax` context value. Please ensure the component is wrapped in a <ParallaxProvider>',
      1
    );
  });

  // fix this
  it.skip('to create an element in the controller on mount', () => {
    const controller = ParallaxController.init({ scrollAxis: VERTICAL });
    controller.createElement = jest.fn();

    render(
      <MockProvider controllerMock={controller}>
        <Parallax y={[-100, 100]}>
          <div />
        </Parallax>
      </MockProvider>
    );

    expect(controller.createElement).toBeCalledWith({
      elInner: (
        <div class="parallax-inner">
          <div />
        </div>
      ),
      elOuter: (
        <div class="parallax-outer">
          <div class="parallax-inner">
            <div />
          </div>
        </div>
      ),
      props: { disabled: false, x0: 0, x1: 0, y0: -100, y1: 100 },
    });
  });

  it('to remove an element in the controller when unmounting', () => {
    const controller = ParallaxController.init({ scrollAxis: VERTICAL });
    controller.removeElementById = jest.fn();

    const { unmount } = render(
      <MockProvider controllerMock={controller}>
        <Parallax>
          <div />
        </Parallax>
      </MockProvider>
    );
    const element = controller.getElements()[0];
    unmount();
    expect(controller.removeElementById).toBeCalledWith(element.id);
  });

  // fix this
  it.skip('to update an element in the controller when receiving relevant new props', () => {
    const controller = ParallaxController.init({ scrollAxis: VERTICAL });
    controller.updateElementPropsById = jest.fn();

    class StateChanger extends React.Component {
      state = { disabled: false };
      render() {
        return <Parallax {...this.state} />;
      }
    }

    let stateInstance;
    render(
      <MockProvider controllerMock={controller}>
        <StateChanger ref={(ref) => (stateInstance = ref)} />
      </MockProvider>
    );

    const testProps = {
      disabled: true,
      x: [100, -100],
      y: [-100, 100],
    };

    // trigger an update
    stateInstance.setState(testProps);
    const element = controller.getElements()[0];

    expect(controller.updateElementPropsById).toBeCalledWith(element.id, {
      disabled: true,
      x0: 100,
      x1: -100,
      y0: -100,
      y1: 100,
    });

    // should not be called again
    stateInstance.setState({
      ...testProps,
      foo: false,
      bar: true,
    });

    expect(controller.updateElementPropsById).toHaveBeenCalledTimes(1);
  });
  // fix this
  it.skip('to reset styles on an element if the disabled prop is true', () => {
    const controller = ParallaxController.init({ scrollAxis: VERTICAL });
    controller.resetElementStyles = jest.fn();

    class StateChanger extends React.Component {
      state = { disabled: false };
      render() {
        return <Parallax {...this.state} />;
      }
    }

    let stateInstance;
    render(
      <MockProvider controllerMock={controller}>
        <StateChanger ref={(ref) => (stateInstance = ref)} />
      </MockProvider>
    );

    stateInstance.setState({ disabled: true });

    expect(controller.resetElementStyles).toBeCalled();
  });
});
