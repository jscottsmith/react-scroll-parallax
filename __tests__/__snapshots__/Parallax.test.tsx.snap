// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect the <Parallax> component to render correctly 1`] = `
<DocumentFragment>
  <figure
    class="parallax-outer class-foo"
    style="border: 2px solid red;"
  >
    <div
      class="parallax-inner"
      style="border: 2px solid blue; transform: translate3d(100%, 100%, 0);"
    >
      <div
        class="foo"
      />
    </div>
  </figure>
</DocumentFragment>
`;
